### PR TITLE
imp: add source credentials

### PR DESCRIPTION
**Pull request by blaggacao**
closes: #18

@nlewo this works on multi-user install, not sure if it works the same way on daemon install or if it depends on a special `nix.conf`, though.

### CODE REVIEW

**Comment by nlewo** [comment]
> this works on multi-user install

Do you mean `mono-user` install?

On multiuser installs, you would need to provide these variables to the deamon. The netrc file is often used for this purpose (see https://nixos.wiki/wiki/Enterprise). The `pullImage` derivation would then have to parse this file (which needs to be bindmounted in the sandbox with `sandbox-paths`).

On NixOS, providing environment variables to the Nix deamon may be tricky because the `nix.conf` file is in the store and we don't want to put secrets in the store. In this context, i think the netrc implementation would be required.

I think we could however start with environment variables, and iterate later when needed.

File: default.nix [hunk 1]
```diff
@@ -80,7 +80,7 @@ let
       dir = pkgs.runCommand name
       {
         inherit imageDigest;
-        impureEnvVars = pkgs.lib.fetchers.proxyImpureEnvVars;
+        impureEnvVars = pkgs.lib.fetchers.proxyImpureEnvVars ++ ["SRC_REGISTRY_USER" "SRC_REGISTRY_PASSWORD"];
```
Comment (nlewo) [inline default.nix:83]: What about adding the `NIX2CONTAINER_` prefix in order to explicit show where this environment variables are used.
```suggestion
        impureEnvVars = pkgs.lib.fetchers.proxyImpureEnvVars ++ ["NIX2CONTAINER_SRC_REGISTRY_USER" "NIX2CONTAINER_SRC_REGISTRY_PASSWORD"];
```

**Comment by blaggacao** [comment]
I'll do a quick&simple bash netrc parser. This is the way to go.

Caveat: will be a single line parser.

**Comment by blaggacao** [comment]
So `netrc` is too low-level, instead let's simply work with the scopeo-native `auth.json` or `$HOME/.docker/config.json`:

```
void builtinFetchurl(const BasicDerivation & drv, const std::string & netrcData)
{
    /* Make the host's netrc data available. Too bad curl requires
       this to be stored in a file. It would be nice if we could just
       pass a pointer to the data. */
    if (netrcData != "") {
        settings.netrcFile = "netrc";
        writeFile(settings.netrcFile, netrcData, 0600);
    }

```

**Comment by blaggacao** [comment]
Refactored & force-pushed and now it works like so:
```
sandbox-paths = /etc/skopeo/auth.json=/home/blaggacao/.docker/config.json
```

**Comment by nlewo** [comment]
@blaggacao neat!
Would it be possible to add a piece of doc  (in the `pullImage`function docstring  for instance)?
Otherwise, lgtm.

**Comment by nlewo** [comment]
Thank you!

**Pull request merged**